llama-index-packs/llama-index-packs-diff-private-examples-gen/llama_index/packs/diff_private_simple_dataset/base.py

```diff
@@ -0,0 +1,294 @@
+                dataset=filtered_simple_dataset,
+                num_splits=num_splits,
+                num_samples_per_split=num_samples_per_split,
+            )
+
+            # generate next token probability distributions per split
+            splits = []
+            for split in disjoint_splits:
+                split_probs = {token: 0 for token in token_universe_probas}
+                prompt = self._get_private_prompt(split, synthetic_example, label)
+                response = self.llm.complete(prompt)
+                # updating and (rescaling) split probs
+                try:
+                    next_token_proba_distn = response.logprobs[0]
+                except IndexError:
+                    print(response)
+                    return FALLBACK_SYNTHETIC_EXAMPLE
+                for el in next_token_proba_distn:  # for immediate next token only
+                    if el.token in split_probs:
+                        split_probs[el.token] = np.exp(el.logprob)
+                split_probs = self._normalize(
+                    split_probs, token_universe_probas
+                )  # to make into a valid prob distribution
+                splits.append(split_probs)
+
+            # noisy aggrergation
+            sigma_calib = np.sqrt(2) / num_splits * sigma
+            noise_array = self._generate_noise(
+                sigma=sigma_calib, size=len(token_universe_probas), mechanism="gaussian"
+            )
+            merged_probas = self._merge_probas(splits)
+            noisy_probs = self._add_noise(merged_probas, noise_array)
+
+            # next token
+            next_token = self._mode_of_distribution(noisy_probs)
+            if next_token in STOP_TOKENS:
+                break
+            else:
+                synthetic_example += next_token
+
+        # synthetic example remove [RESULT]
+        try:
+            synthetic_example = synthetic_example.split("[RESULT] ")[-1]
+        except Exception as e:
+            synthetic_example = synthetic_example
+
+        return LabelledSimpleDataExample(
+            reference_label=label,
+            text=synthetic_example,
+            text_by=CreatedBy(type=CreatedByType.AI, model_name=self.llm.model),
+        )
+
+    def run(
+        self,
+        sizes: Dict[str, int],
+        t_max: int = 1,
+        sigma: float = 0.5,
+        num_splits: int = 5,
+        num_samples_per_split: int = 1,
+    ) -> LabelledSimpleDataset:
+        """Main run method."""
+        if num_samples_per_split < 1:
+            raise ValueError(
+                "`num_samples_per_split` must be an integer greater than 1."
+            )
+
+        if not all(c in sizes for c in self._labels):
+            raise ValueError("Not all labels have sizes.")
+
+        examples = []
+        for label in self._labels:
+            size = sizes[label]
+            for _ in range(size):
+                example = self.generate_dp_synthetic_example(
```

Comment:
if you wanted this to be async, this top-level function is where I would do it. The actually dp generation algorithm seems fairly linear, but if it was using async LLM calls, this top level could be run concurrently ?

Reply:
ah yeah so this loop invokes the generate synthetic example as you know.... within this method i think there is room to do more async:

`generate_dp_synthetic_example` does something like this:
 ```
 def generate_dp_synthetic_example():
         synthetic_example = ""
         for _ in max_iterations:
               split_distributions = {}
               for split in splits:
                     # generate candidate next-token (rather their proba distribution) per split
                     next_token_proba_distn = get_next_token_probab_from_llm(split, synthetic_example)
                     
                     split_distributions[split] = next_token_proba_distn
               
               merged_distribution = noisy_aggregation(split_distributions)
               
               next_token = argmax(merged_distribution)
               synthetic_example += next_token
 ```
 
 ☝️ pseudo-code, names of methods differ from actual code just for the sake of demonstration. I think we can do async stuff at the split level as well — that is get the split_distributions via a asyncio.gather()?